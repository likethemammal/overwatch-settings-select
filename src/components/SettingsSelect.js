import React, { Component } from 'react'
import _ from 'underscore'

import Menu  from 'material-ui/Menu'

import DownSVG from 'svg-icon/dist/svg/mfglabs/chevron_down.svg'

import general from '@likethemammal/general-components'


const { SVG } = general.components

import css from './SettingsSelect.less'

import {
    selectedOption as _selectedOption,
} from '../selectors/select'

const menuOverrideProps = {
    marginThreshold: 0,
    getContentAnchorEl: null,
    anchorOrigin: {
        vertical: 'bottom',
        horizontal: 'left',
    },
    MenuListProps: {
        disablePadding: true,
        className: css.list_override,
        classes: {
            root: css.list_root_override,
        }
    },
    PaperProps: {
        style: {
            background: 'transparent'
        }
    }
}

export default class SettingsSelect extends Component {

    state = {
        open: false,
        anchorEl: null,
    }

    onMenuOpen = (ev) => {
        this.setState({ open: true, anchorEl: ev.currentTarget })

        ev.target.blur()
    }

    onMenuClose = (ev) => {
        this.setState({ open: false, })
    }

    onSelect = (index) => {
        this.setState({ open: false })

        this.props.onChange(index)
    }

    render() {
        const { options, label, selectedIndex } = this.props

        const {
            open,
            anchorEl,
        } = this.state

        const selectedOption = _selectedOption(this.props)

        return <div className={css.container}>
            <div className={css.inner}>

                <div
                    className={!open ? css.label : css.label__selected}
                >
                    {label}
                </div>

                <div className={css.select}>

                    <button
                        onClick={this.onMenuOpen}
                        className={css.active}
                    >
                        {selectedOption}
                        <div className={css.icon_container}>
                            <SVG
                                raw={DownSVG}
                                className={css.icon}
                            />
                        </div>
                    </button>

                    <Menu
                        {...menuOverrideProps}
                        id={label}
                        open={open}
                        anchorEl={anchorEl}
                        onBackdropClick={this.onMenuClose}
                    >
                        {_.map(options, (option, i) => {
                            return <li
                                onClick={this.onSelect.bind(null, i)}
                                key={i}
                                className={css.list_item}
                                selected={i === selectedIndex}
                            >
                                {option}
                            </li>
                        })}
                    </Menu>
                </div>
            </div>
        </div>
    }
}
