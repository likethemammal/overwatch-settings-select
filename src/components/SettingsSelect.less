@import "../../node_modules/@likethemammal/overwatch-general/lib/colors.less";

@item_width: 377px;

.container {
    &:hover {
        .label {
            background: @gray_faded_setting_hover;
        }
    }

    //low browser support, so cant be combined with :hover selector
    &:focus-within {
        .label {
            background: @gray_faded_setting_hover;
        }
    }
}

.inner {
    position: relative;
    padding-right: @item_width + 7px;
}

.__text {
    composes: _font_general from '@likethemammal/overwatch-general/lib/general.css';
    height: 4.3rem;
    line-height: 4.4rem;
    color: hsl(0, 0%, 90%);
}

.label {
    composes: __text;
    composes: cf from '@likethemammal/general-components/bin/general.css';
    padding-left: 22px;
    background: @black_faded_setting;
    transition: background 100ms ease-in;
    font-size: 2.3rem;
}

.label__selected {
    composes: label;
    background: @gray_faded_setting_hover;
}

.select {
    position: absolute;
    top: 0;
    right: 0;
}

.list_override {
    padding-right: 0 !important;
}

.list_root_override {
    width: 100% !important;
}

.item {
    composes: __text;
    width: @item_width;
    line-height: 4.5rem;
    padding-left: 20px;
    letter-spacing: 0.01rem;
    font-size: 2.2rem;
    cursor: pointer;
}

.list_item {
    composes: item;
    background: @blue_solid;

    &:focus {
        border: 0;
        outline: 0;
        background: @blue_solid_hover_dark;
    }

    &:hover {
        background: @blue_solid_hover_dark;
    }
}

.active {
    composes: item;
    background: @blue_faded_hover;
    border: 0;
    text-align: left;

    &:focus {
        border: 0;
        outline: 0;
    }
}

.icon_container {
    composes: _icon_container from '@likethemammal/overwatch-general/lib/general.css';
    position: absolute;
    right: 5px;
    left: auto;
    top: -1px;
}
