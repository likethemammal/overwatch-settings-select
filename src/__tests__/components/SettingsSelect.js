import React from 'react';
import {shallow} from 'enzyme';

import general from '@likethemammal/general-components'

const shared = general.__tests__._shared.tests

import SettingsSelect from '../../../lib'

const options = ['At the sk', 'fucksd sd', 'Some option']
const label = 'Some label'
const selectedIndex = 2
const onChange = jest.fn()

const props = {
    options,
    label,
    selectedIndex,
    onChange,
}


describe('SettingsSelect', () => {
    // Render a checkbox with label in the document
    const component = shallow(
        <SettingsSelect
            {...props}
        />
    )

    shared.SNAPSHOTS_SHOULD_MATCH([
        component
    ])

})