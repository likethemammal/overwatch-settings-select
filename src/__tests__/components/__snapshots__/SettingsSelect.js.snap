// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsSelect Snapshot 0 should match 1`] = `undefined`;
